File: frontend/flutter_app/lib/services/favorito_service.dart
import '../models/receta_favorita.dart';
import 'api_client.dart';

/// Servicio HTTP para la gestión de recetas favoritas.
class FavoritoService {
  /// POST /api/favoritos — Marcar una receta como favorita.
  static Future<RecetaFavorita> marcarFavorito({
    required int recetaId,
    required String token,
  }) async {
    final body = {'recetaId': recetaId};

    final data = await ApiClient.post('/favoritos', body, token: token);
    return RecetaFavorita.fromJson(data as Map<String, dynamic>);
  }

  /// DELETE /api/favoritos — Desmarcar una receta como favorita.
  static Future<void> desmarcarFavorito({
    required int recetaId,
    required String token,
  }) async {
    await ApiClient.delete('/favoritos?recetaId=$recetaId', token: token);
  }

  /// GET /api/favoritos/usuario — Obtener favoritos del usuario autenticado.
  static Future<List<RecetaFavorita>> obtenerFavoritos({
    required String token,
  }) async {
    final data = await ApiClient.get('/favoritos/usuario', token: token);
    return (data as List)
        .map((e) => RecetaFavorita.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  /// GET /api/favoritos/verificar — Verificar si una receta es favorita.
  static Future<bool> verificarFavorito({
    required int recetaId,
    required String token,
  }) async {
    final data = await ApiClient.get(
      '/favoritos/verificar?recetaId=$recetaId',
      token: token,
    );
    return data['esFavorita'] as bool;
  }

  /// GET /api/favoritos/usuario/ids — IDs de favoritos del usuario autenticado.
  static Future<List<int>> obtenerIdsFavoritos({required String token}) async {
    final data = await ApiClient.get('/favoritos/usuario/ids', token: token);
    return (data as List).map((e) => e as int).toList();
  }

  /// Alternar favorito: marca si no lo está, desmarca si lo está.
  static Future<bool> toggleFavorito({
    required int recetaId,
    required bool esFavoritaActualmente,
    required String token,
  }) async {
    if (esFavoritaActualmente) {
      await desmarcarFavorito(recetaId: recetaId, token: token);
      return false;
    } else {
      await marcarFavorito(recetaId: recetaId, token: token);
      return true;
    }
  }
}

File: backend/src/main/java/com/ares/backend/service/FavoritoService.java
package com.ares.backend.service;

import com.ares.backend.config.SecurityUtils;
import com.ares.backend.dto.FavoritoRequest;
import com.ares.backend.dto.RecetaDetailResponse;
import com.ares.backend.dto.RecetaFavoritaResponse;
import com.ares.backend.entity.Receta;
import com.ares.backend.entity.RecetaFavorita;
import com.ares.backend.entity.Usuario;
import com.ares.backend.repository.RecetaFavoritaRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Servicio para la gestión de favoritos.
 *
 * @author Ares
 * @version 1.0
 */
@Service
@RequiredArgsConstructor
public class FavoritoService {

    private final RecetaFavoritaRepository favoritoRepository;
    private final UsuarioService usuarioService;
    private final RecetaService recetaService;

    /**
     * Marca una receta como favorita para un usuario.
     *
     * @param request Datos del favorito (usuarioId, recetaId)
     * @return Favorito creado
     * @throws IllegalArgumentException Si ya existe el favorito
     */
    @Transactional
    public RecetaFavoritaResponse marcarFavorito(FavoritoRequest request) {
        Long usuarioId = SecurityUtils.getUsuarioId();
        Usuario usuario = usuarioService.buscarPorId(usuarioId);
        Receta receta = recetaService.buscarPorId(request.getRecetaId());

        // Verificar si ya existe
        if (favoritoRepository.existsByUsuarioAndReceta(usuario, receta)) {
            throw new IllegalArgumentException("La receta ya está marcada como favorita");
        }

        // Crear favorito
        RecetaFavorita favorito = new RecetaFavorita(usuario, receta);
        favorito = favoritoRepository.save(favorito);

        // Construir respuesta
        RecetaDetailResponse recetaDetail = recetaService.obtenerPorId(receta.getId());
        recetaDetail.setEsFavorita(true);

        return new RecetaFavoritaResponse(favorito, recetaDetail);
    }

    /**
     * Desmarca una receta como favorita.
     *
     * @param request Datos del favorito (usuarioId, recetaId)
     * @throws IllegalArgumentException Si el favorito no existe
     */
    @Transactional
    public void desmarcarFavorito(FavoritoRequest request) {
        Long usuarioId = SecurityUtils.getUsuarioId();
        Usuario usuario = usuarioService.buscarPorId(usuarioId);
        Receta receta = recetaService.buscarPorId(request.getRecetaId());

        // Verificar si existe
        if (!favoritoRepository.existsByUsuarioAndReceta(usuario, receta)) {
            throw new IllegalArgumentException("La receta no está marcada como favorita");
        }

        // Eliminar favorito
        favoritoRepository.deleteByUsuarioAndReceta(usuario, receta);
    }

    /**
     * Obtiene todas las recetas favoritas del usuario autenticado.
     *
     * @return Lista de recetas favoritas con sus detalles
     */
    public List<RecetaFavoritaResponse> obtenerFavoritosUsuario() {
        Long usuarioId = SecurityUtils.getUsuarioId();
        Usuario usuario = usuarioService.buscarPorId(usuarioId);
        List<RecetaFavorita> favoritos = favoritoRepository.findByUsuarioOrderByFechaMarcadoDesc(usuario);

        return favoritos.stream()
                .map(favorito -> {
                    RecetaDetailResponse recetaDetail = recetaService.obtenerPorId(favorito.getReceta().getId());
                    recetaDetail.setEsFavorita(true);
                    return new RecetaFavoritaResponse(favorito, recetaDetail);
                })
                .collect(Collectors.toList());
    }

    /**
     * Verifica si una receta es favorita del usuario autenticado.
     *
     * @param recetaId ID de la receta
     * @return true si es favorita, false en caso contrario
     */
    public boolean esFavorita(Long recetaId) {
        Long usuarioId = SecurityUtils.getUsuarioId();
        Usuario usuario = usuarioService.buscarPorId(usuarioId);
        Receta receta = recetaService.buscarPorId(recetaId);
        return favoritoRepository.existsByUsuarioAndReceta(usuario, receta);
    }

    /**
     * Obtiene los IDs de todas las recetas favoritas del usuario autenticado.
     * Útil para marcar favoritos en listados.
     *
     * @return Lista de IDs de recetas favoritas
     */
    public List<Long> obtenerIdsFavoritosUsuario() {
        Long usuarioId = SecurityUtils.getUsuarioId();
        return favoritoRepository.findRecetaIdsByUsuarioId(usuarioId);
    }
}